clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java

```diff
@@ -106,8 +381,15 @@ public synchronized void setPidAndEpoch(long pid, short epoch) {
      * In all of these cases, we don't know whether batch was actually committed on the broker, and hence whether the
      * sequence number was actually updated. If we don't reset the producer state, we risk the chance that all future
      * messages will return an OutOfOrderSequenceException.
+     *
+     * Note that we can't reset the producer state for the transactional producer as this would mean bumping the epoch
+     * for the same pid. This might involve aborting the ongoing transaction during the initPidRequest, and the user
+     * would not have any way of knowing this happened. So for the transactional producer, it's best to return the
+     * produce error to the user and let them abort the transaction and close the producer explicitly.
      */
     public synchronized void resetProducerId() {
+        if (isTransactional())
+            return;
```

Comment:
Silent `return` here seems a bit dangerous. Maybe we should throw an exception and require the caller to check  `isTransactional` itself.